grafana/resource_datasource_permission.go

```diff
@@ -0,0 +1,229 @@
+				Description: "The permission items to add/update. Items that are omitted from the list will be removed.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"team_id": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "ID of the team to manage permissions for.",
+						},
+						"user_id": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "ID of the user to manage permissions for.",
+						},
+						"permission": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.StringInSlice([]string{"Query"}, false),
+							Description:  "Permission to associate with item. Must be `Query`.",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func UpdateDatasourcePermissions(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client).gapi
+
+	v, ok := d.GetOk("permissions")
+	if !ok {
+		return nil
+	}
+	datasourceID := int64(d.Get("datasource_id").(int))
+
+	configuredPermissions := []*gapi.DatasourcePermissionAddPayload{}
+	for _, permission := range v.(*schema.Set).List() {
+		permission := permission.(map[string]interface{})
+		permissionItem := gapi.DatasourcePermissionAddPayload{}
+		if permission["team_id"].(int) != -1 {
+			permissionItem.TeamID = int64(permission["team_id"].(int))
+		}
+		if permission["user_id"].(int) != -1 {
+			permissionItem.UserID = int64(permission["user_id"].(int))
+		}
+		permissionItem.Permission = mapDatasourcePermissionStringToInt64(permission["permission"].(string))
+		configuredPermissions = append(configuredPermissions, &permissionItem)
+	}
+
+	if err := updateDatasourcePermissions(client, datasourceID, configuredPermissions, true, false); err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(strconv.FormatInt(datasourceID, 10))
+
+	return ReadDatasourcePermissions(ctx, d, meta)
+}
+
+func ReadDatasourcePermissions(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client).gapi
+
+	id, err := strconv.ParseInt(d.Id(), 10, 64)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	response, err := client.DatasourcePermissions(id)
+	if err != nil {
+		if strings.HasPrefix(err.Error(), "status: 404") {
+			log.Printf("[WARN] removing datasource permissions %d from state because it no longer exists in grafana", id)
+			d.SetId("")
+			return nil
+		}
+
+		return diag.FromErr(err)
+	}
+
+	permissionItems := make([]interface{}, len(response.Permissions))
+	count := 0
+	for _, permission := range response.Permissions {
+		permissionItem := make(map[string]interface{})
+		permissionItem["team_id"] = permission.TeamID
+		permissionItem["user_id"] = permission.UserID
+		permissionItem["permission"] = mapDatasourcePermissionInt64ToString(permission.Permission)
+
+		permissionItems[count] = permissionItem
+		count++
+	}
```

Comment:
There's no need for the `count` variable, you can make use of the index returned by `range`:
```suggestion
	permissionItems := make([]interface{}, len(response.Permissions))
	for i, permission := range response.Permissions {
		permissionItem := make(map[string]interface{})
		permissionItem["team_id"] = permission.TeamID
		permissionItem["user_id"] = permission.UserID
		permissionItem["permission"] = mapDatasourcePermissionInt64ToString(permission.Permission)

		permissionItems[i] = permissionItem
	}
```

Reply:
ah, oops. This is what I get for copy pasting